cluster/log-dump/log-dump.sh

```diff
@@ -0,0 +1,363 @@
+    wait
+  fi
+}
+
+function dump_nodes_with_logexporter() {
+  echo "Detecting nodes in the cluster"
+  detect-node-names &> /dev/null
+
+  if [[ "${#NODE_NAMES[@]}" == 0 ]]; then
+    echo "No nodes found!"
+    return
+  fi
+
+  # Obtain parameters required by logexporter.
+  local -r service_account_credentials="$(cat ${GOOGLE_APPLICATION_CREDENTIALS} | base64)"
+  local -r cloud_provider="${KUBERNETES_PROVIDER}"
+  local -r gcs_artifacts_dir="${GCS_ARTIFACTS_DIR}"
+  local -r enable_hollow_node_logs="${ENABLE_HOLLOW_NODE_LOGS:-false}"
+  local -r logexport_timeout_seconds="$(( 30 + NUM_NODES / 10 ))"
+
+  # Fill in the parameters in the logexporter daemonset template.
+  sed -i'' -e "s/{{.ServiceAccountCredentials}}/${service_account_credentials}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.CloudProvider}}/${cloud_provider}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.GCSPath}}/${gcs_artifacts_dir}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.EnableHollowNodeLogs}}/${enable_hollow_node_logs}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+
+  # Create the logexporter namespace, service-account secret and the logexporter daemonset within that namespace.
+  KUBECTL="${KUBECTL:-${KUBE_ROOT}/cluster/kubectl.sh}"
+  "${KUBECTL}" create -f "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+
+  # Give some time for the pods to finish uploading logs.
+  sleep "${logexport_sleep_seconds}"
+
+  # List the logexporter pods created and their corresponding nodes.
+  pods_and_nodes=()
+  for retry in {1..5}; do
+    pods_and_nodes=$(${KUBECTL} get pods -n logexporter -o=custom-columns=NAME:.metadata.name,NODE:.spec.nodeName | tail -n +2)
+    if [[ -n "${pods_and_nodes}" ]]; then
+      echo -e "List of logexporter pods found:\n${pods_and_nodes}"
+      break
+    fi
+    if [[ "${retry}" == 5 ]]; then
+      echo "Failed to list any logexporter pods after multiple retries.. falling back to logdump for nodes through SSH"
+      "${KUBECTL}" delete namespace logexporter
+      dump_nodes "${NODE_NAMES[@]}"
+      return
+    fi
+  done
+
+  # Collect names of nodes we didn't find a logexporter pod on.
+  # Note: This step is O(#nodes^2) as we check if each node is present in the list of nodes running logexporter.
+  # Making it linear would add code complexity without much benefit (as it just takes < 1s for 5k nodes anyway).
+  failed_nodes=()
+  for node in "${NODE_NAMES[@]}"; do
+    if [[ ! "${pods_and_nodes}" =~ "${node}" ]]; then
+      failed_nodes+=("${node}")
+    fi
+  done
+
+  # Collect names of nodes whose logexporter pod didn't succeed.
+  # TODO(shyamjvs): Parallelize the for loop below to make it faster (if needed).
+  logexporter_pods=( $(echo "${pods_and_nodes}" | awk '{print $1}') )
+  logexporter_nodes=( $(echo "${pods_and_nodes}" | awk '{print $2}') )
+  for index in "${!logexporter_pods[@]}"; do
```

Comment:
It is a convenient way of iterating through 0-(length-1). Otherwise the expression would look ugly as we have to calculate length-1.